pkg/internal/source/kind.go

```diff
@@ -104,14 +131,61 @@ func (ks *Kind) String() string {
 // WaitForSync implements SyncingSource to allow controllers to wait with starting
 // workers until the cache is synced.
 func (ks *Kind) WaitForSync(ctx context.Context) error {
+	if func() bool {
+		ks.mu.Lock()
+		defer ks.mu.Unlock()
+
+		return ks.started == nil
+	}() {
+		return fmt.Errorf("cannot wait for sync on an unstarted source")
+	}
+
 	select {
 	case err := <-ks.started:
+		if err != nil {
+			// Try to stop the source to clean up any resources
+			// and collect any errors.
+			if stopErr := ks.Stop(); stopErr != nil {
+				err = kerrors.NewAggregate([]error{err, fmt.Errorf("failed to stop source: %w", stopErr)})
+			}
+		}
 		return err
```

Comment:
nit: can we maybe return nil here and return err at the end of the if branch? Just to make it slightly more obvious that this here is also a non-error case